draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -3774,32 +3772,92 @@ Input: u, an element of F.
 Output: (x, y), a point on E.
 
 Constants:
-1.  c1 = -B / A
-2.  c2 = -1 / Z
+1.  c1 = Z / S
 
 Steps:
-1.  tv1 = Z * u^2
-2.  tv2 = tv1^2
-3.   x1 = tv1 + tv2
-4.   x1 = inv0(x1)
-5.   e1 = x1 == 0
-6.   x1 = x1 + 1
-7.   x1 = CMOV(x1, c2, e1)    # If (tv1 + tv2) == 0, set x1 = -1 / Z
-8.   x1 = x1 * c1 # x1 = (-B / A) * (1 + (1 / (Z^2 * u^4 + Z * u^2)))
-9.  gx1 = x1^2
-10. gx1 = gx1 + A
-11. gx1 = gx1 * x1
-12. gx1 = gx1 + B             # gx1 = g(x1) = x1^3 + A * x1 + B
-13.  x2 = tv1 * x1            # x2 = Z * u^2 * x1
-14. tv2 = tv1 * tv2
-15. gx2 = gx1 * tv2           # gx2 = (Z * u^2)^3 * gx1
-16.  e2 = is_square(gx1)      # If is_square(gx1)
-17.   x = CMOV(x2, x1, e2)    #   then  x = x1,  else  x = x2
-18.  y2 = CMOV(gx2, gx1, e2)  #   then y2 = gx1, else y2 = gx2
-19.   y = sqrt(y2)
-20.  e3 = sgn0(u) == sgn0(y)  # Fix sign of y
-21.   y = CMOV(-y, y, e3)
-22. return (x, y)
+1.  tv1 = u^2
+2.  tv1 = Z * tv1
+3.  tv2 = tv1^2
+4.  tv2 = tv2 + tv1
+5.  tv3 = tv2 + 1
+6.  tv3 = B * tv3
+7.  tv4 = CMOV(Z, -tv2, tv2 != 0)
+8.  tv4 = A * tv4
+9.  tv2 = tv3^2
+10. tv5 = tv4^2
+11. tv5 = A * tv5
+12. tv2 = tv2 + tv5
+13. tv2 = tv2 * tv3
+14. tv6 = tv4^3
```

Comment:
The RHS here is x1<sub>div</sub><sup>3</sup> in the notation used in the Zcash protocol spec (https://zips.z.cash/protocol/protocol.pdf#concretegrouphashpallasandvesta). We previously computed x1<sub>div</sub><sup>3</sup> as tv5 in step 10, and so there is no need to do a cubing here; a multiplication of (tv5 in step 10) * tv4 would be sufficient.

In general, I think that reusing temporary variables is unhelpful: it's more error-prone and harder for humans to understand, while the first thing a compiler is going to do anyway (regardless of platform, and platforms with few registers are not an exception) is to convert it to static single-assignment form.

Reply:
Got the fix in the new PR.

I agree regarding temporary variables, but I'm going to focus on local fixes in my PR.

Reply:
Temporaries here are just the same as we do elsewhere in the doc. If we make a change here (which would be fine), we should at least be consistent.